lib/components/joystick/joystick_action.dart

```diff
@@ -0,0 +1,256 @@
+    switch (align) {
+      case JoystickActionAlign.TOP_LEFT:
+        dx = margin.left + radius;
+        dy = margin.top + radius;
+        break;
+      case JoystickActionAlign.BOTTOM_LEFT:
+        dx = margin.left + radius;
+        dy = _screenSize.height - (margin.bottom + radius);
+        break;
+      case JoystickActionAlign.TOP_RIGHT:
+        dx = _screenSize.width - (margin.right + radius);
+        dy = margin.top + radius;
+        break;
+      case JoystickActionAlign.BOTTOM_RIGHT:
+        dx = _screenSize.width - (margin.right + radius);
+        dy = _screenSize.height - (margin.bottom + radius);
+        break;
+    }
+    _rectAction = Rect.fromCircle(
+      center: Offset(dx, dy),
+      radius: radius,
+    );
+    _rectBackgroundDirection = Rect.fromCircle(
+      center: Offset(dx, dy),
+      radius: _sizeBackgroundDirection / 2,
+    );
+
+    if (spriteBackgroundDirection == null)
+      _paintBackground = Paint()
+        ..color = color.withOpacity(opacityBackground)
+        ..style = PaintingStyle.fill;
+
+    if (sprite == null)
+      _paintAction = Paint()
+        ..color = color.withOpacity(opacityKnob)
+        ..style = PaintingStyle.fill;
+
+    if (spritePressed == null)
+      _paintActionPressed = Paint()
+        ..color = color.withOpacity(opacityBackground)
+        ..style = PaintingStyle.fill;
+
+    _dragPosition = _rectAction.center;
+  }
+
+  void render(Canvas c) {
+    if (_rectBackgroundDirection != null && _dragging && enableDirection) {
+      if (spriteBackgroundDirection == null) {
+        final double radiusBackground = _rectBackgroundDirection.width / 2;
+        c.drawCircle(
+          Offset(
+            _rectBackgroundDirection.left + radiusBackground,
+            _rectBackgroundDirection.top + radiusBackground,
+          ),
+          radiusBackground,
+          _paintBackground,
+        );
+      } else {
+        spriteBackgroundDirection.renderRect(c, _rectBackgroundDirection);
+      }
+    }
+
+    if (_spriteAction != null) {
+      if (_rectAction != null) {
+        _spriteAction.renderRect(c, _rectAction);
+      }
+    } else {
+      final double radiusAction = _rectAction.width / 2;
+      c.drawCircle(
+        Offset(
+          _rectAction.left + radiusAction,
+          _rectAction.top + radiusAction,
+        ),
+        radiusAction,
+        isPressed ? _paintActionPressed : _paintAction,
+      );
+    }
+  }
+
+  void update(double dt) {
+    if (_dragging) {
+      final double _radAngle = atan2(
+        _dragPosition.dy - _rectBackgroundDirection.center.dy,
+        _dragPosition.dx - _rectBackgroundDirection.center.dx,
+      );
+
+      // Distance between the center of joystick background & drag position
+      final Point centerPoint = Point(
+        _rectBackgroundDirection.center.dx,
+        _rectBackgroundDirection.center.dy,
+      );
+      double dist =
+          centerPoint.distanceTo(Point(_dragPosition.dx, _dragPosition.dy));
```

Comment:
any reason to use Point here instead of Position? position provides distance between two points and you can convert from offsets (Position.fromOffset(_rectBackgroundDirection.center))